pkg/git/v2/transport.go

```diff
@@ -48,35 +55,66 @@ func AuthSecretStrategyForURL(URL string) (common.AuthSecretStrategy, error) {
 
 type BasicAuth struct{}
 
-func (s *BasicAuth) Method(secret corev1.Secret) (*common.Auth, error) {
-	var username string
-	if d, ok := secret.Data["username"]; ok {
-		username = string(d)
-	}
-	var password string
-	if d, ok := secret.Data["password"]; ok {
-		password = string(d)
-	}
-	if username == "" || password == "" {
-		return nil, fmt.Errorf("invalid '%s' secret data: required fields 'username' and 'password'", secret.Name)
+func (s *BasicAuth) Method(secret *corev1.Secret, certSecret *corev1.Secret) (*common.Auth, error) {
+	var credCallback git2go.CredentialsCallback = nil
+	if secret != nil {
+		var username string
+		if d, ok := secret.Data["username"]; ok {
+			username = string(d)
+		}
+		var password string
+		if d, ok := secret.Data["password"]; ok {
+			password = string(d)
+		}
+		if username == "" || password == "" {
+			return nil, fmt.Errorf("invalid '%s' secret data: required fields 'username' and 'password'", secret.Name)
+		}
+		credCallback = func(url string, username_from_url string, allowed_types git2go.CredType) (*git2go.Cred, error) {
+			cred, err := git2go.NewCredUserpassPlaintext(username, password)
+			if err != nil {
+				return nil, err
+			}
+			return cred, nil
+		}
 	}
 
-	credCallback := func(url string, username_from_url string, allowed_types git2go.CredType) (*git2go.Cred, error) {
-		cred, err := git2go.NewCredUserpassPlaintext(username, password)
-		if err != nil {
-			return nil, err
+	var certCallback git2go.CertificateCheckCallback = nil
+	if certSecret != nil {
+		caCert, ok := certSecret.Data[CACert]
+		if !ok {
+			return nil, fmt.Errorf("invalid '%s' cert secret data: required field '%s'", secret.Name, CACert)
+		}
+
+		certCallback = func(cert *git2go.Certificate, valid bool, hostname string) git2go.ErrorCode {
```

Comment:
The valid parameter is true if the system cert validation passed. @stefanprodan should we ignore the valid parameter and always do our own validation or skip the validation if it is true?

Reply:
If a CA is provided, I think we should run the validation no matter what the system check returns.